### PR TITLE
log errors on make_encoder()

### PR DESCRIPTION
**Pull request by patrick-prima**
We have an odd problem where a small schema update is causing avrogen to fail on startup. We think a little more logging might help us understand what is going on.

For reference, these PRs fail:
* https://github.com/primait/uk-quote-request-transformers/pull/29
* https://github.com/primait/uk-quote-request-transformers/pull/34

```
** (Mix) Could not start application uk_avro_schema: UkAvroSchema.Application.start(:normal, []) returned an error: shutdown: failed to start child: Avrogen.Schema.SchemaRegistry
    ** (EXIT) an exception was raised:
        ** (ErlangError) Erlang error: "union should have at least one member type"
            (erlavro 2.9.10) /Users/xxx/projects/uk-quote-request-transformers/deps/erlavro/src/avro_union.erl:73: :avro_union.type/1
            (erlavro 2.9.10) /Users/xxx/projects/uk-quote-request-transformers/deps/erlavro/src/avro_json_decoder.erl:72: :avro_json_decoder.decode_schema/2
            (erlavro 2.9.10) /Users/xxx/projects/uk-quote-request-transformers/deps/erlavro/src/avro_util.erl:77: :avro_util.ensure_lkup_fun/1
            (erlavro 2.9.10) /Users/xxx/projects/uk-quote-request-transformers/deps/erlavro/src/avro.erl:208: :avro.make_encoder/2
            (avrogen 0.6.6) lib/avrogen/schema/schema_registry.ex:43: anonymous fn/1 in Avrogen.Schema.SchemaRegistry.init/1
            (noether 0.2.5) lib/noether/either.ex:23: Noether.Either.map/2
            (avrogen 0.6.6) lib/avrogen/schema/schema_registry.ex:41: Avrogen.Schema.SchemaRegistry.init/1
            (stdlib 4.3.1.3) gen_server.erl:851: :gen_server.init_it/2
```

### CODE REVIEW

File: lib/avrogen/schema/schema_registry.ex [hunk 1]
```diff
@@ -5,6 +5,7 @@ defmodule Avrogen.Schema.SchemaRegistry do
   """
 
   use GenServer
+  use Logger
```
Comment (oxg558) [inline lib/avrogen/schema/schema_registry.ex:8]: I think `require Logger` is more usual

File: lib/avrogen/schema/schema_registry.ex [hunk 2]
```diff
@@ -39,10 +40,23 @@ defmodule Avrogen.Schema.SchemaRegistry do
...
+        json = Jason.encode!(schemas)
+        encoder = make_encoder(json)
+        decoder = make_decoder(json)
+        :ets.insert(__MODULE__, {@ets_name, json, encoder, decoder})
+      rescue
+        e ->
+          formatted_error = Exception.format(:error, e, __STACKTRACE__)
+
+          Logger.error("Error when attempting to make encoder/decoder.",
+            error: formatted_error,
```
Comment (oxg558) [inline lib/avrogen/schema/schema_registry.ex:53]: will we be able to see the logger metadata in the CI? I think we might only see the main text

Comment (patrick-prima) [reply lib/avrogen/schema/schema_registry.ex:53]: I think you get the metadata if your hipster-level is over 9000 and you click on sufficiently many links, buttons and icons, but I might be wrong.

I can add more plain text.

File: lib/avrogen/schema/schema_registry.ex (hunk 1 above)
Comment (patrick-prima) [reply lib/avrogen/schema/schema_registry.ex:8]: Ah yes.

**Review by oxg558 — DISMISSED** [review]
LGTM! Would be a lot easier to debug if we could recreate the problem locally...

**Comment by oxg558** [comment]
Though since the error itself happens quite deep in an erlavro call I'm not sure if this log will tell us much more... it will at least show us what order the schemas are being sorted in, which might be what has changed.

**Review by oxg558 — APPROVED** [review]

File: lib/avrogen/schema/schema_registry.ex [hunk 3]
```diff
@@ -39,10 +40,27 @@ defmodule Avrogen.Schema.SchemaRegistry do
...
-      :ets.insert(__MODULE__, {@ets_name, json, encoder, decoder})
+      try do
+        json = Jason.encode!(schemas)
+        encoder = make_encoder(json)
+        decoder = make_decoder(json)
+        :ets.insert(__MODULE__, {@ets_name, json, encoder, decoder})
+      rescue
+        e ->
+          formatted_error = Exception.format(:error, e, __STACKTRACE__)
+          json = Jason.encode(schemas) |> elem(1)
```
Comment (oxg558) [inline lib/avrogen/schema/schema_registry.ex:51]: maybe just put the original `json = Jason.encode!(schemas)` outside the try and access it here

File: lib/avrogen/schema/schema_registry.ex [hunk 4]
```diff
@@ -39,10 +40,27 @@ defmodule Avrogen.Schema.SchemaRegistry do
...
+
+          Logger.error(
+            "Error when attempting to make encoder/decoder: #{formatted_error}; schemas: #{inspect(schemas)} schemas_json: #{json}",
```
Comment (oxg558) [inline lib/avrogen/schema/schema_registry.ex:54]: `schemas` is very long so might get truncated by `inspect`.

`IO.inspect(schemas, limit: :infinity, printable_limit: :infinity)`

**Review by oxg558 — APPROVED** [review]

File: lib/avrogen/schema/schema_registry.ex [hunk 5]
```diff
@@ -40,9 +41,26 @@ defmodule Avrogen.Schema.SchemaRegistry do
...
+      try do
+        encoder = make_encoder(json)
+        decoder = make_decoder(json)
+        :ets.insert(__MODULE__, {@ets_name, json, encoder, decoder})
+      rescue
+        e ->
+          formatted_error = Exception.format(:error, e, __STACKTRACE__)
+
+          Logger.error(
+            "Error when attempting to make encoder/decoder: #{formatted_error}; schemas: #{inspect(schemas, limit: :infinity, pretty: true, printable_limit: :infinity)} schemas_json: #{elem(json, 1)}",
```
[the remark this answers is not in the document]
Comment (patrick-prima) [reply lib/avrogen/schema/schema_registry.ex]: Of course...

**Review by oxg558 — DISMISSED** [review]

**Comment by patrick-prima** [comment]
We found the culprit.

**Pull request closed without merging**